sw/device/tests/i2c_host_ambient_light_detector_test.c

```diff
@@ -0,0 +1,153 @@
+enum {
+  kDeviceAddr = 0x29,
+
+  // Registers
+  kPartIdReg = 0x86,
+  kManufacturerIdReg = 0x87,
+  kAlsContrReg = 0x80,
+  kAlsMeasRateReg = 0x85,
+  kAlsDataCh10Reg = 0x88,
+  kAlsDataCh11Reg = 0x89,
+  kAlsDataCh00Reg = 0x8A,
+  kAlsDataCh01Reg = 0x8B,
+  kAlsStatusReg = 0x8C,
+
+  // Registers values
+  kPartId = 0xA0,
+  kManufacturerId = 0x05,
+  kCtrlActive = 0x01,
+  kMeasRate = 0x01 << 3 | 0x02,
+
+  // Register masks
+  kDataStatus = 1 << 2,
+};
+
+static dif_rv_core_ibex_t rv_core_ibex;
+static dif_pinmux_t pinmux;
+static dif_i2c_t i2c;
+
+static status_t read_part_id(void) {
+  uint8_t reg = kPartIdReg, data = 0;
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, 1, &reg, true));
+  TRY(i2c_testutils_read(&i2c, kDeviceAddr, 1, &data));
+  TRY_CHECK(data == kPartId, "Unexpected value %x", data);
+  return OK_STATUS();
+}
+
+static status_t read_manufacturer_id(void) {
+  uint8_t reg = kManufacturerIdReg, data = 0;
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, 1, &reg, true));
+  TRY(i2c_testutils_read(&i2c, kDeviceAddr, 1, &data));
+  TRY_CHECK(data == kManufacturerId, "Unexpected value %x", data);
+  return OK_STATUS();
+}
+
+static status_t take_measurement(void) {
+  // Set the measurement rate to something non-default.
+  uint8_t meas_reg = kAlsMeasRateReg;
+  uint8_t rate[2] = {meas_reg, kMeasRate};
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, sizeof(rate), rate, true));
+  // Read back to confirm the write.
+  uint8_t rate_data = 0;
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, 1, &meas_reg, true));
+  TRY(i2c_testutils_read(&i2c, kDeviceAddr, 1, &rate_data));
+  TRY_CHECK(rate_data == kMeasRate, "Unexpected value %x", rate_data);
+
+  // Enable active measurements.
+  uint8_t ctrl_reg = kAlsContrReg;
+  uint8_t ctrl[2] = {ctrl_reg, kCtrlActive};
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, sizeof(ctrl), ctrl, true));
+  // Read back the ctrl register to confirm active measurement mode was set.
+  uint8_t ctrl_data = 0;
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, 1, &ctrl_reg, true));
+  TRY(i2c_testutils_read(&i2c, kDeviceAddr, 1, &ctrl_data));
+  TRY_CHECK(ctrl_data == kCtrlActive, "Unexpected value %x", ctrl_data);
+
+  // Poll until the status register sets the "data status" bit.
+  uint8_t status_reg = kAlsStatusReg;
+  uint8_t status = 0;
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, 1, &status_reg, true));
+  do {
+    TRY(i2c_testutils_read(&i2c, kDeviceAddr, sizeof(status), &status));
+  } while ((status & kDataStatus) == 0);
+
+  // Read data in the order it's laid out in memory: 10, 11, 00, 01.
+  uint8_t data_start = kAlsDataCh10Reg;
+  uint8_t data[4] = {0x00, 0x00, 0x00, 0x00};
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, 1, &data_start, true));
+  TRY(i2c_testutils_read(&i2c, kDeviceAddr, sizeof(data), data));
+
+  // Check data isn't all 0x00 or all 0xFF, which are more likely to be errors
+  // than real values.
+  uint8_t all_ff[4] = {0xFF, 0xFF, 0xFF, 0xFF};
+  uint8_t all_00[4] = {0x00, 0x00, 0x00, 0x00};
```

Comment:
I think it's worth printing the value read here.

Reply:
Added a `LOG_INFO`. FYI, they're printed when the `TRY_CHECK_ARRAYS_NE` fails, too